src/include/homestore/blkdata_service.hpp

```diff
@@ -224,6 +226,19 @@ class BlkDataService {
     std::unique_ptr< BlkReadTracker > m_blk_read_tracker;
     std::shared_ptr< ChunkSelector > m_custom_chunk_selector;
     uint32_t m_blk_size;
+
+private:
+    // graceful shutdown related
+    std::atomic_bool m_stopping{false};
+    mutable std::atomic_uint64_t pending_request_num{0};
```

Comment:
Can we use a [Counting Semaphore](https://en.cppreference.com/w/cpp/thread/counting_semaphore) so we can make it event driven instead of check->sleep->check logic?

Reply:
thanks for the suggestion, I got your point here. well , I think the scenario here is a little different from the one suitable for counting_semaphore.

1 counting_semaphore is used to set a limitation to the max concurrency of actions simultaneously. in the scenario here, I do not want set any limitation. all the api can be called as much as possible as long as the service is not stopped.

2 when starting shutting down, we should exactly know whether all the pending requests are completed. however , counting_semaphore does not provide an api to see if all the action that acquired the semaphore has release it.  with counting_semaphore, what we can do it only to acquire or try_acquire.

pls correct me if I am wrong @szmyd

Reply:
You're right; I am wrong, I had it inverted in my head. What we'd want is a shared_mutex that everyone takes a read lock (increment the lock count), and the "waiter" takes a write_lock which will only unlock once the readers are done.

This solution you have currently is fine since it only is exercised during shutdown though. I will continue to review and ignore this part.